src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+            Contract.ThrowIfNull(oldSemanticTokensData);
+
+            var edits = new SemanticTokensEdits
+            {
+                Edits = ComputeSemanticTokensEdits(oldSemanticTokensData, newSemanticTokensData),
+                ResultId = resultId
+            };
+
+            return edits;
+        }
+
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdit[] ComputeSemanticTokensEdits(
+            int[] oldSemanticTokens,
+            int[] newSemanticTokens)
+        {
+            // We use Roslyn's version of the Myers' Diff Algorithm to compute the minimal edits
+            // between the old and new tokens.
+            // Edits are computed by token (i.e. in sets of five integers), so if one value in the token
+            // is changed, the entire token is replaced. We do this instead of directly comparing each
+            // value in the token individually so that we can potentially save on computation costs, since
+            // we can return early if we find that one value in the token doesn't match. However, there
+            // are trade-offs since our insertions/deletions are usually larger.
+
+            // Turning arrays into tuples of five ints, each representing one token
+            var oldGroupedSemanticTokens = ConvertToGroupedSemanticTokens(oldSemanticTokens);
+            var newGroupedSemanticTokens = ConvertToGroupedSemanticTokens(newSemanticTokens);
+
+            var edits = LongestCommonSemanticTokensSubsequence.GetEdits(oldGroupedSemanticTokens, newGroupedSemanticTokens);
+
+            return ConvertToSemanticTokenEdits(newGroupedSemanticTokens, edits);
+        }
+
+        private static SemanticTokensEdit[] ConvertToSemanticTokenEdits(SemanticToken[] newGroupedSemanticTokens, IEnumerable<SequenceEdit> edits)
+        {
+            // Our goal is to minimize the number of edits we return to LSP. It's possible an index
+            // may have both an insertion and deletion, in which case we can combine the two into a
+            // single update. We use the dictionary below to keep track of whether an index contains
+            // an insertion, deletion, or both.
+            var indexToEditKinds = new Dictionary<int, SemanticTokenEditKind>();
+
+            foreach (var edit in edits)
+            {
+                // We only care about EditKind.Insert and EditKind.Delete, since they encompass all
+                // changes to the document. All other EditKinds are ignored.
+                switch (edit.Kind)
+                {
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert |= SemanticTokenEditKind.Insert;
```

Comment:
interesting.  so SmenaticTokenEdit "ors" teh values together?  i.e. you could have an edit at an index that is both an insert and a delete?  do you have tests for this?

Reply:
Yep, the semantic token edits tests have scenarios covering all three possible cases (insert, delete, and update).